### PR TITLE
Switch Interaction Model client error reporting to just CHIP_ERROR.

### PR DESCRIPTION
**Pull request by bzbarsky-apple**
Now that we can put a StatusIB inside a CHIP_ERROR, we can stop
passing both, or passing just EmberAfStatus, and pass a CHIP_ERROR
that either contains a StatusIB or contains the actual client-side
error we ran into (e.g. failure to decode).

#### Problem
We are trying to force a StatusIB through an EmberAfStatus, it's not clear to consumers how the CHIP_ERROR and the StatusIB (or EmberAfStatus or InteractionModel::Status) interact when both are passed, etc.

#### Change overview
Switch all the IM failure/error callbacks to take CHIP_ERROR.

#### Testing
Should be no behavior changes for now.  The consumers that could really use this (in CHIPDeviceController, where we are ending up treating a general status as a cluster-specific status) are not using the new APIs yet.

### CODE REVIEW

File: src/controller/TypedReadCallback.h [hunk 1]
```diff
@@ -28,22 +28,24 @@ namespace chip {
...
+ * This provides an adapter class that implements ReadClient::Callback and provides three additional
  * features:
  *  1. The ability to pass in std::function closures to permit more flexible programming scenarios than are provided by the strict
  *     delegate interface stipulated by by InteractionModelDelegate.
  *
  *  2. Automatic decoding of attribute data provided in the TLVReader by InteractionModelDelegate::OnReportData into a decoded
  *     cluster object.
+ *
+ *  3. Automatically representing all errors as a CHIP_ERROR (which might
+ *     encapsulate a StatusIB).
```
Comment (mrjerryjohns) [inline src/controller/TypedReadCallback.h:40]: Unify comment please.

Comment (bzbarsky-apple) [reply src/controller/TypedReadCallback.h]: Done.

File: src/app/WriteClient.h [hunk 2]
```diff
@@ -223,10 +224,9 @@ class WriteClient : public Messaging::ExchangeDelegate
     // Timed Request.  The caller is assumed to have already checked that our
     // exchange context member is the one the message came in on.
     //
-    // aStatusIB will be populated with the returned status if we can parse it
-    // successfully.
-    CHIP_ERROR HandleTimedStatus(const PayloadHeader & aPayloadHeader, System::PacketBufferHandle && aPayload,
-                                 StatusIB & aStatusIB);
+    // If the server returned an error status, that will be returned as an error
```
[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply src/app/WriteClient.h]: Done.

File: src/app/TimedRequest.h [hunk 3]
```diff
@@ -36,8 +36,10 @@ class TimedRequest
 
     // Handle a response message (which may not actually be a StatusResponse,
     // but came in after we sent a timed request).
-    static CHIP_ERROR HandleResponse(const PayloadHeader & aPayloadHeader, System::PacketBufferHandle && aPayload,
-                                     StatusIB & aStatusIB);
+    //
+    // If the response is a failure StatusResponse, its status will be returned
```
[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply src/app/TimedRequest.h]: Done.

File: src/app/CommandSender.h [hunk 4]
```diff
@@ -279,10 +277,9 @@ class CommandSender final : public Messaging::ExchangeDelegate
     // Timed Request.  The caller is assumed to have already checked that our
     // exchange context member is the one the message came in on.
     //
-    // aStatusIB will be populated with the returned status if we can parse it
-    // successfully.
-    CHIP_ERROR HandleTimedStatus(const PayloadHeader & aPayloadHeader, System::PacketBufferHandle && aPayload,
-                                 StatusIB & aStatusIB);
+    // If the server returned an error status, that will be returned as an error
```
[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply src/app/CommandSender.h:282]: Done.

**Review by jmeg-sfy — APPROVED** [review]

**Review by msandstedt — APPROVED** [review]

**Review by andy31415 — APPROVED** [review]

**Review by woody-apple — APPROVED** [review]

**Pull request merged**